src/bin/pg_autoctl/cli_do_tmux.c

```diff
@@ -246,117 +276,185 @@ tmux_add_send_keys_command(PQExpBuffer script, const char *fmt, ...)
 
 
 /*
- * tmux_add_xdg_environment appends the XDG environment that makes the test
- * target self-contained, as a series of tmux send-keys commands, to the given
- * script buffer.
+ * tmux_prepare_XDG_environment set XDG environment variables in the current
+ * process tree.
+ */
+static bool
+tmux_prepare_XDG_environment(const char *root, bool createDirectories)
+{
+	log_info("Preparing XDG setting for self-contained session in \"%s\"",
+			 root);
+
+	for (int i = 0; xdg[i][0] != NULL; i++)
+	{
+		char *var = xdg[i][0];
+		char *dir = xdg[i][1];
+		char *env = (char *) malloc(MAXPGPATH * sizeof(char));
+
+		if (env == NULL)
+		{
+			log_fatal("Failed to malloc MAXPGPATH bytes: %m");
+			return false;
+		}
+
+		sformat(env, MAXPGPATH, "%s/%s", root, dir);
+
+		if (createDirectories)
+		{
+			log_debug("mkdir -p \"%s\"", env);
+			if (pg_mkdir_p(env, 0700) == -1)
+			{
+				log_error("mkdir -p \"%s\": %m", env);
+				return false;
+			}
+		}
+
+		if (!normalize_filename(env, env, MAXPGPATH))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		log_info("export %s=\"%s\"", var, env);
+
+		if (setenv(var, env, 1) != 0)
+		{
+			log_error("Failed to set environment variable %s to \"%s\": %m",
+					  var, env);
+		}
+
+		/* also create our actual target directory for our files */
+		if (createDirectories)
+		{
+			char targetPath[MAXPGPATH] = { 0 };
+
+			sformat(targetPath, sizeof(targetPath),
+					"%s/pg_config/%s", env, root);
+
+			log_debug("mkdir -p \"%s\"", targetPath);
+			if (pg_mkdir_p(targetPath, 0700) == -1)
+			{
+				log_error("mkdir -p \"%s\": %m", targetPath);
+				return false;
+			}
+		}
+	}
+
+	return true;
+}
+
+
+/*
+ * tmux_add_new_session appends a new-session command with the
+ * update-environment options for our XDG settings, as a series of tmux
+ * send-keys commands, to the given script buffer.
  */
 static void
-tmux_add_xdg_environment(PQExpBuffer script, const char *root)
+tmux_add_new_session(PQExpBuffer script, const char *root, int pgport)
 {
-	char *xdg[][3] = {
-		{ "XDG_DATA_HOME", "share" },
-		{ "XDG_CONFIG_HOME", "config" },
-		{ "XDG_RUNTIME_DIR", "run" },
-	};
+	char sessionName[BUFSIZE] = { 0 };
+
+	sformat(sessionName, BUFSIZE, "pgautofailover-%d", pgport);
 
 	/*
 	 * For demo/tests purposes, arrange a self-contained setup where everything
 	 * is to be found in the given options.root directory.
 	 */
-	for (int i = 0; i < 3; i++)
+	for (int i = 0; xdg[i][0] != NULL; i++)
 	{
 		char *var = xdg[i][0];
-		char *dir = xdg[i][1];
 
-		tmux_add_send_keys_command(script,
-								   "export %s=\"%s/%s\"", var, root, dir);
+		tmux_add_command(script, "set-option update-environment %s", var);
```

Comment:
This still doesn't fix my issue where a second cluster has the first clusters XDG env vars. After a little googling I think we should change this to (non pseudo code version of):
```suggestion
		tmux_add_command(script, "setenv %s %s", var, normalize_path(xdg[i][1]));
```

Reply:
So the `set-option update-environment` does not work as I though it would. I didn't try the `setenv` command, I have added back the sending of the `export var=value` commands as we used to have before. I could then use `pg_autoctl do tmux session` on 3 different environments in parallel.